### PR TITLE
Add a serve_stale option for plugin/cache

### PR DESCRIPTION
**Pull request by gonzalop**
### 1. Why is this pull request needed and what does it do?
Implements #3158 by adding a _serve_stale_ option to the cache directive.
From unbound's man page:
>  serve-expired: <yes or no>
>               If  enabled,  unbound attempts to serve old responses from cache with a TTL of 0 in
>               the response without waiting for the  actual  resolution  to  finish.   The  actual
>               resolution answer ends up in the cache later on.  Default is "no".

### 2. Which issues (if any) are related?
#3158 

### 3. Which documentation changes (if any) need to be made?
Yes, the PR includes a documentation update.

### 4. Does this introduce a backward incompatible change or deprecation?
No.

### CODE REVIEW

File: plugin/cache/handler.go
Comment (chrisohaver) [inline plugin/cache/handler.go:44]: IIUC, this go routine is the "ends up in the cache later on" bit.
Will this cause CoreDNS to send a DNS response packet to itself?

Comment (gonzalop) [reply plugin/cache/handler.go:44]: No, nothing is written to the response when prefetch is true.
The LocalAddr is arbitrary and is only needed to avoid a nullref in  ResponseWriter.RemoteAddr (it is not called, but I did not want to take any chances). Once a backend returns a cacheable response, it is cached but never written.

File: plugin/cache/README.md
Comment (chrisohaver) [inline plugin/cache/README.md]: "... when there are no healthy upstream backends."

IIUC, we are not literally watching the health of the upstream backends here.  So this may be misleading.  One might infer that if an expired entry is served to the client, that must mean that the backend/upstream is unhealthy/non-responsive - but that is not always the case.  The actual behavior may be harder to succinctly describe, but IMO we should try...

When serve_expired is set, CoreDNS always will serve an expired cache to a client if an expired cache entry exists.  When this is happens, CoreDNS will attempt to refresh the cache entry after sending the expired cache entry to the client.

Comment (gonzalop) [reply plugin/cache/README.md]: You are correct.
I've updated  the text with your more accurate description. Let me know if there's anything that looks off.

Thanks

-Gonzalo

**Review by chrisohaver — APPROVED** [review]
Thanks for the contribution! LGTM. Although IMO we should get an approval from one of the plugin OWNERS before merging ...

**Comment by gonzalop** [comment]
Ack. Thank you, Chris, for reviewing.

File: plugin/cache/README.md
Comment (miekg) [inline plugin/cache/README.md]: s/CoreDNS/*cache*/

File: plugin/cache/README.md
Comment (miekg) [inline plugin/cache/README.md]: I would remove the hint, sounds a bit pedantic

File: plugin/cache/README.md
Comment (miekg) [inline plugin/cache/README.md]: the new metrics should also be added to the README

File: plugin/cache/handler.go
Comment (miekg) [inline plugin/cache/handler.go]: why do we need to reget the item; we've already looked it up. It's it more natural to have `c.get` return an boolean that includes that's it expired?
Or maybe a new function getExpired, but we control the intree code so we can upgrade all callers.

File: plugin/cache/setup_test.go
Comment (miekg) [inline plugin/cache/setup_test.go]: just put these checks in a small new test that just checks the new elements. Now we just have a giant list of `false` and a few `true`'s that check the new setup code. A new selfcontained small test is almost certainly better.

**Review by miekg — CHANGES_REQUESTED** [review]

File: plugin/cache/README.md
Comment (stp-ip) [reply plugin/cache/README.md]: Also: This be true|false instead of yes|no as far as I can see.

Comment (stp-ip) [reply plugin/cache/README.md]: Couldn't we assume serve_expired to be false by default and therefore not need the true|false at all?
This would simplify the config to 'serve_expired [duration]'
Also generally called 'stale' responses instead of expired.

File: plugin/cache/README.md
Comment (gonzalop) [reply plugin/cache/README.md]: Done.

File: plugin/cache/README.md
Comment (gonzalop) [reply plugin/cache/README.md]: This is something I would have never thought of without external feedback, but you are right. Removed. Thanks.

File: plugin/cache/README.md
Comment (gonzalop) [reply plugin/cache/README.md]: I've added the metrics. renamed the option to serve_stale and got rid of the yes|no argument.
Thanks! :)

File: plugin/cache/handler.go
Comment (gonzalop) [reply plugin/cache/handler.go]: You are right.
I added a getItem to unconditionally get an item from the cache. PTAL.

File: plugin/cache/setup_test.go
Comment (gonzalop) [reply plugin/cache/setup_test.go]: Done. A lot cleaner now.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add a serve_e..." ]
>stp-ip commented on this pull request.
>
>
>
>> @@ -34,6 +34,7 @@ cache [TTL] [ZONES...] {
>     success CAPACITY [TTL] [MINTTL]
>     denial CAPACITY [TTL] [MINTTL]
>     prefetch AMOUNT [[DURATION] [PERCENTAGE%]]
>+    serve_expired yes|no [DURATION]
>
>Also: This be true|false instead of yes|no as far as I can see.

or just the presence of the string `serve_expired`, so the boolean is 
superfluous.

File: plugin/cache/README.md
Comment (miekg) [inline plugin/cache/README.md]: `0h` or did you mean `1h` here?

File: plugin/cache/cache.go
Comment (miekg) [inline plugin/cache/cache.go]: these two values can be unified into 1: if staleUpTo is zero it is disabled, otherwise we use it.

File: plugin/cache/handler.go
Comment (miekg) [inline plugin/cache/handler.go]: getIgnoreTTL or something? getItem is very simular to get. Might also benefit from some documentation :)

File: plugin/cache/setup.go
Comment (miekg) [inline plugin/cache/setup.go]: "invalid negative duration for serve_stale"

File: plugin/cache/setup.go
Comment (miekg) [inline plugin/cache/setup.go]: just return `err` as in the other parseDurations here.

File: plugin/cache/setup.go
Comment (miekg) [inline plugin/cache/setup.go]: return return nil, c.ArgErr()

**Review by miekg — COMMENTED** [review]
looks much better, thanks, still some (style) comments though

File: plugin/cache/README.md
Comment (gonzalop) [reply plugin/cache/README.md]: I'm removing that. There is no default now and you have to specify the duration.

File: plugin/cache/cache.go
Comment (gonzalop) [reply plugin/cache/cache.go]: How would you know when to serve immediately + fetch vs the normal behavior when the options is not present?

File: plugin/cache/README.md
Comment (chrisohaver) [reply plugin/cache/README.md]: Per setup.go, looks like duration is optional and default is 1h.

File: plugin/cache/setup.go
Comment (gonzalop) [reply plugin/cache/setup.go]: Done. Thanks.

File: plugin/cache/setup.go
Comment (gonzalop) [reply plugin/cache/setup.go]: done

File: plugin/cache/handler.go
Comment (gonzalop) [reply plugin/cache/handler.go]: Done.

...but I noticed the 2 metrics in the old get method would now never be incremented. I am now incrementing those metrics as well, PTAL.

File: plugin/cache/setup.go
Comment (gonzalop) [reply plugin/cache/setup.go]: done

File: plugin/cache/README.md
Comment (gonzalop) [reply plugin/cache/README.md]: Yup, it's updated in my last push.

File: plugin/cache/cache.go
Comment (miekg) [reply plugin/cache/cache.go]: isn't `serveStale == staleUpTo.IsZero()` ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add a serve_s..." ]
>gonzalop commented on this pull request.
>
>
>
>> @@ -83,6 +99,18 @@ func (c *Cache) get(now time.Time, state request.Request, server string) (*item,
> 	return nil, false
> }
>
>+func (c *Cache) getItem(state request.Request, server string) *item {
>
>I noticed the 2 metrics in the old get method would now never be incremented. Only getIgnoredTTL knows where the item came from (Denial, Success). What would be the right thing to do here?

file an issue, please? Leave it for now - bug fixing and features should be kept 
in separate PRs if possible.

File: plugin/cache/cache.go
Comment (gonzalop) [reply plugin/cache/cache.go]: If a configuration does not include serve_stale, we would have enabled the feature for ttl == 0.

Comment (gonzalop) [reply plugin/cache/cache.go]: It breaks a prefetch test that expects to trigger a fetch after 30s for a TTL=30.

Comment (gonzalop) [reply plugin/cache/cache.go]: Done.
I got waaaaaay derailed by one of my tests but in the end you were right and it was trivial to remove it. Thanks.

**Comment by gonzalop** [comment]
Anything else I should do?

**Comment by miekg** [comment]
I'm trying out the new and improved corbot (github.com/miekg/dreck)

/lgtm
/merge

**Comment by miekg** [comment]
hmmm the lgtm led to a 
~~~
[ERROR] dreck: POST https://api.github.com/repos/coredns/coredns/pulls/3468/reviews: 422 Unprocessable Entity [{Resource: Field: Code: Message:}]
~~~
which is REST for 'no' ?

**Pull request merged**
